api/controllers/v0.3/ILSHelper.js

```diff
@@ -0,0 +1,390 @@
+        // something like "Dorothy" or "Katherine " or " Jackson".
+        // We'll take what we can get.
+        fullName = firstName || lastName;
+      }
+    }
+
+    let fields = {
+      name: fullName,
+      address: address,
+      username: patron.username,
+      pin: patron.pin,
+      ptype: patron.ptype,
+      workAddress: workAddress,
+      ecommunications_pref: patron.ecommunicationsPref,
+    };
+
+    if (patron.barcode) {
+      fields["barcode"] = patron.barcode;
+    }
+    if (email) {
+      fields["email"] = email;
+    }
+    if (birthdate) {
+      fields["birthdate"] = birthdate;
+    }
+
+    return fields;
+  }
+
+  // Determine whether a patron should have a permanent card.
+  longExpiration(patron) {
+    // False if a patron's existing work address isn't commercial
+    if (patron.workAddress && patron.workAddress.isResidential) {
+      return false;
+    }
+
+    // False if patron provides a home address that is not residential
+    // False if patron does not have a recognized name
+    // False if patron policy is not the default (:simplye)
+    return (
+      patron.address.isResidential &&
+      patron.hasValidName &&
+      patron.policy.isDefault
+    );
+  }
+
+  patronFields(patron) {
+    let fieldData = this.formattedPatronData(patron);
+
+    let fields = [
+      fieldData["name"],
+      fieldData["address"],
+      fieldData["ptype"],
+      fieldData["username"],
+      fieldData["pin"],
+      fieldData["ecommunications_pref"],
+      // notice_pref_as_field,
+      // home_lib_as_field,
+      patron.policy.agency,
+    ];
+
+    if (patron.worksInCity) {
+      fields.push(fieldData["workAddress"]);
+      fields.push(note_as_field);
+    }
+
+    if (this.longExpiration(patron)) {
+      fields.push(this.longExpirationAsField(patron));
+    } else {
+      patron.set_temporary();
+      fields.push(this.shortExpirationAsField(patron));
+    }
+
+    // Add required/optional fields according to policy.
+    if (patron.barcode) {
+      fields.push(fieldData["barcode"]);
+    }
+    if (patron.email) {
+      fields.push(fieldData["email"]);
+    }
+    if (patron.birthdate) {
+      fields.push(fieldData["birthdate"]);
+    }
+
+    return {
+      patronFields: {
+        patron_field: fields,
+      },
+      patronID: patron.patronId,
+    };
+  }
+
+  blankField() {
+    return { fieldTag: "", marcTag: 0, value: "", value_is_binary: false };
+  }
+
+  customField(fieldTag, value, marcTag = 0) {
+    let field = this.blankField();
+    field["field_tag"] = fieldTag;
+    field["value"] = value;
+    field["marc_tag"] = marcTag;
+    return field;
+  }
+
+  expirationAsField(time) {
+    // TODO convert date object
+    let expiration = time.strftime("%Y%m%d000000");
+    return this.customField(
+      IlsHelper.EXPIRATION_FIELDTAG,
+      expiration,
+      IlsHelper.STRING_MARCTAG
+    );
+  }
+
+  longExpirationAsField(patron) {
```

Comment:
Is there a reason to keep this and `shortExpirationAsField` as separate functions (instead of passing an argument for standard/temporary)?

Reply:
The previous code had a lot of "every function should only do one specific thing". Is that normal ruby convention as well? In any case, yes, there's lot of optimizations we can do but still reading the code and trying to make sure that's possible to do. There are some cases where many functions aren't even used, and some cases where margin will be harder than expected.

Reply:
Atomic functions are normal Ruby convention, but...this is pretty extreme.